src/main/java/uk/gov/justice/hmpps/prison/service/OffenderEventsService.java

```diff
@@ -0,0 +1,94 @@
+package uk.gov.justice.hmpps.prison.service;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import uk.gov.justice.hmpps.prison.api.model.OffenderEvent;
+import uk.gov.justice.hmpps.prison.api.resource.impl.OffenderEventsController;
+import uk.gov.justice.hmpps.prison.repository.jpa.repository.OffenderEventsRepository;
+import uk.gov.justice.hmpps.prison.service.filters.OffenderEventsFilter;
+import uk.gov.justice.hmpps.prison.service.transformers.OffenderEventsTransformer;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Service
+@Transactional(readOnly = true)
+public class OffenderEventsService {
+
+    private static final Comparator<uk.gov.justice.hmpps.prison.api.model.OffenderEvent> BY_OFFENDER_EVENT_TIMESTAMP =
+            Comparator.comparing(uk.gov.justice.hmpps.prison.api.model.OffenderEvent::getEventDatetime);
+
+    private static final Comparator<uk.gov.justice.hmpps.prison.api.model.OffenderEvent> BY_OFFENDER_EVENT_TIMESTAMP_DESC = BY_OFFENDER_EVENT_TIMESTAMP.reversed();
+
+    private final OffenderEventsTransformer offenderEventsTransformer;
+    private final OffenderEventsRepository offenderEventsRepository;
+    private final XtagEventsService xtagEventsService;
+
+    @Autowired
+    public OffenderEventsService(final OffenderEventsTransformer offenderEventsTransformer,
+                                 final OffenderEventsRepository offenderEventsRepository,
+                                 final XtagEventsService xtagEventsService) {
+        this.offenderEventsTransformer = offenderEventsTransformer;
+        this.offenderEventsRepository = offenderEventsRepository;
+        this.xtagEventsService = xtagEventsService;
+    }
+
+    @PreAuthorize("hasRole('PRISON_OFFENDER_EVENTS')")
+    public Optional<List<OffenderEvent>> getEvents(final Optional<LocalDateTime> maybeFrom,
```

Comment:
I bet you had IntelliJ telling you not to use Optionals as arguments ...

Reply:
it is